### PR TITLE
Xudong's a5

### PR DESCRIPTION
**Pull request by XudongJia**

### CODE REVIEW

File: Problem1/src/problem1/Problem1.java [hunk 1]
```diff
@@ -25,7 +25,7 @@ public void looping1(int n) {
 
     /**
      * looping2
-     * RUNTIME COMPLEXITY IS: O(?)
+     * RUNTIME COMPLEXITY IS: O(2n)
```
Comment (cgallinaro) [inline Problem1/src/problem1/Problem1.java]: We drop the scalar.  The complexity is simply O(n).

File: Problem1/src/problem1/Problem1.java [hunk 2]
```diff
@@ -64,8 +64,8 @@ public void looping4(int n) {
     }
 
     /**
-     * looping4
-     * RUNTIME COMPLEXITY IS: O(?)
+     * looping5
+     * RUNTIME COMPLEXITY IS: O(n^n)
```
Comment (cgallinaro) [inline Problem1/src/problem1/Problem1.java]: The complexity is O(n^2).

File: Problem1/src/problem1/Problem1.java [hunk 3]
```diff
@@ -95,7 +95,7 @@ public void looping6(int n) {
 
     /**
      * recursive1
-     * RUNTIME COMPLEXITY IS: O(?)
+     * RUNTIME COMPLEXITY IS: O(1)
```
Comment (cgallinaro) [inline Problem1/src/problem1/Problem1.java]: Since the value of n decreases by one on every recursive call, the complexity is O(n).

File: Problem1/src/problem1/Problem1.java [hunk 4]
```diff
@@ -110,7 +110,7 @@ public int recursive1(int n) {
 
     /**
      * recursive2
-     * RUNTIME COMPLEXITY IS: O(?)
+     * RUNTIME COMPLEXITY IS: O(1)
```
Comment (cgallinaro) [inline Problem1/src/problem1/Problem1.java]: This is also O(n).

File: Problem1/src/problem1/Problem1.java [hunk 5]
```diff
@@ -140,7 +140,7 @@ public int recursive3(int n) {
 
     /**
      * recursive4
-     * RUNTIME COMPLEXITY IS: O(?)
+     * RUNTIME COMPLEXITY IS: O(n!)
```
Comment (cgallinaro) [inline Problem1/src/problem1/Problem1.java]: Since every non-base case call results in 2 additional recursive calls, the complexity is O(2^n).

File: Problem2/test/Problem2/Problem2Test.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    @Test
+    public void testChange() {
+        System.out.println("change");
+        String srxx = "123456";
+        long srl = Problem2.change(Long.parseLong(srxx));
+        System.out.println(srl);
+        if (srl % 3 == 0) {
+            System.out.println(srxx + "Could divisible by 3!");
+        } else {
+            System.out.println(srxx + "Couldn't divisible by 3!");
```
Comment (cgallinaro) [inline Problem2/test/Problem2/Problem2Test.java]: You are missing an assertion statement in this test.  You do not actually validate that your method outputs the correct value.

File: Problem2/test/Problem2/Problem2Test.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        }
+    }
+
+}
```
Comment (cgallinaro) [inline Problem2/test/Problem2/Problem2Test.java]: When testing recursive methods, you should ensure that your base cases are covered (i.e., verify that inputs of 3, 6 and 9 return the correct value).

File: Problem2/src/problem2/Problem2.java [hunk 8]
```diff
@@ -5,17 +5,70 @@
...
+        String str = "" + a;
+        char[] xx = str.toCharArray();
+        long b = 0;
+        for (int i = 0; i < xx.length; i++) {
+            b = b + Long.parseLong(xx[i] + "");
+        }
+        if (b > 9) {
+            return change(b);
+        } else {
+            return b;
```
Comment (cgallinaro) [inline Problem2/src/problem2/Problem2.java]: Instead of returning the value, this method should be checking the base cases (i.e. check if b equals 3, 6 or 9).

File: Problem3/src/problem3/Problem3.java [hunk 9]
```diff
@@ -5,17 +5,30 @@
...
     public static void main(String[] args) {
-        // TODO code application logic here
+        
+        Person[] persons = new Person[5];
+	persons[0] = new Person("10004", "one", 20);
+	persons[1] = new Person("10002", "two", 18);
+	persons[2] = new Person("10003", "three", 19);
+	persons[3] = new Person("10001", "four", 20);
+	persons[4] = new Person("10005", "four", 20);
+	Arrays.sort(persons);//排序
```
Comment (cgallinaro) [inline Problem3/src/problem3/Problem3.java]: The question asked you to implement a version of insertion sort.  You were not supposed to use the Arrays.sort method.

File: Problem3/test/problem3/Problem3Test.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    /**
+     * Test of main method, of class Problem3.
+     */
+    @Test
+    public void testMain() {
+        System.out.println("main");
+        String[] args = null;
+        Problem3.main(args);
```
Comment (cgallinaro) [inline Problem3/test/problem3/Problem3Test.java]: If you had implemented a version of insertion sort, you would have been able to implement a few test cases for it.

File: Problem4/src/problem4/Problem4.java [hunk 11]
```diff
@@ -5,17 +5,86 @@
...
+    public static boolean isNumeric(String str) {
+        Pattern pattern = Pattern.compile("[0-9]*");
+        Matcher isNum = pattern.matcher(str);
+        return isNum.matches();
+    }
+
+    //bubble sort
+    /**
+     * insertion sort :O(N^2/2) 5,2,4,6,1,3 
+     * My algorithm:O(3/4*N^2)15,20,36,12,45,9
```
Comment (cgallinaro) [inline Problem4/src/problem4/Problem4.java]: Since the scalar values are ignored, your algorithm also has a complexity of O(n^2).  Also, you forgot to answer the following:
Compare this algorithm to insertion sort
- describe an array which would perform better using your algorithm (over insertion sort)
- describe an array which would perform better using insertion sort (over your algorithm)

File: Problem4/test/problem4/Problem4Test.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+     * Test of sort method, of class Problem4.
+     */
+    @Test
+    public void testSort() {
+        System.out.println("sort");
+        String str = "1,3,2,4";
+        System.out.println("Before："+str);
+        System.out.println("After："+Problem4.sort(str));
```
Comment (cgallinaro) [inline Problem4/test/problem4/Problem4Test.java]: You are missing an assertion statement in this test.

File: Problem4/test/problem4/Problem4Test.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    }
+}
```
Comment (cgallinaro) [inline Problem4/test/problem4/Problem4Test.java]: When testing sorting algorithms, you should ensure you have test coverage for edge cases:
- Empty array
- Array of one element

**Comment by cgallinaro** [comment]
Mark:

Problem 1: 3/5
Problem 2: 4/5
Problem 3: 2.5/5
Problem 4: 8/10

Total: 17.5/25
